### PR TITLE
[#81] Update the project to use the up-to-date versions

### PR DESCRIPTION
**Pull request by hoangmirs**
Resolves https://github.com/nimblehq/nimble-survey-web/issues/81

## What happened 👀

- Upgrade to Ruby to 3.0.6 - we had an issue when running the latest one due to the `figaro` gem
- Upgrade to Rails to 7
- Upgrade to Node 18
- Drop Webpacker, change to use `esbuild` for JS and `sass`, `postcss` for CSS
- Replace `scss-lint` with `stylelint`
- Add `sprockets-rails` to load files in Rails
- Introduce `Makefile` and update README

## Insight 📝

- Rails 7 doesn't support Webpacker, which was why we need to use other tools
- Add `sprockets-rails` back to compile gem stylesheets, and also it helps load other files (images, fonts, SVGs)
- Move all `devDependencies` to `dependencies` to use them in Dockerfile
- Add `.yarnrc` in `.yarnrc` to allow install the current `@nimblehq/stylelint-config-nimble` as it doesn't support the latest node.
- Create the `Makefile` to reduce the step to develop. We also refactor files to make the development faster.
- In `make test`, we use the different Docker containers with dev as isolating the container is better IMO. And it can help us in case we didn't start the dev containers, then it could use the test containers separately.
- Add `volumes` to `docker-compose.dev.yml` to persist data on localhost.

## Proof Of Work 📹

Deployed and tested on Staging:

||Screenshots|
|----|--------|
|OAuth applications|Doorkeeper|
|Forgot password API|Forgot_password_-_Nimble|
|Survey list API|Survey_list_-_Nimble|
|GraphQL endpoint|GraphqQL__sample__-_Nimble|
|Email sent|Spark_-_hoang_n_nimblehq_co|
|Reset password screen|Nimble_Survey|

### CODE REVIEW

**Comment by hoangmirs** [comment]
copilot:walkthrough

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-14.17.0
+18.15.0
```
Comment (suho) [inline .nvmrc:1]: As we are using asdf `.tool-versions`, do we need to keep this? 🤔

Comment (hoangmirs) [reply .nvmrc]: Good catch, removed all unused files in https://github.com/nimblehq/nimble-survey-web/pull/84/commits/37ee301bd357e6629e41ce48dda596bdd13b4f93

**Review by suho — APPROVED** [review]

File: .github/workflows/danger.yml [hunk 2]
```diff
@@ -20,7 +20,7 @@ jobs:
       - name: Setup Ruby
         uses: ruby/setup-ruby@v1
         with:
-          ruby-version: "2.7.1"
+          ruby-version: "3.0.6"
```
Comment (andyduong1920) [inline .github/workflows/danger.yml:23]: do we need to upgrade the node version below

Comment (hoangmirs) [reply .github/workflows/danger.yml:23]: Good catch. Updated in https://github.com/nimblehq/nimble-survey-web/pull/84/commits/00f35f6803d8f58b3d244f1e36c99255f28393f9

File: app/controllers/api/v1/surveys_controller.rb [hunk 3]
```diff
@@ -27,7 +27,7 @@ def show
 
       def pagination_params
         {
-          page: params.dig(:page, :number) || Pagy::VARS[:page],
+          page: params.dig(:page, :number),
```
[the remark this answers is not in the document]
Comment (hoangmirs) [reply app/controllers/api/v1/surveys_controller.rb:30]: It seesm `Page::VARS` is not available, then I had to change it in `config/initializers.pagy.rb`. And that was why I removed this too.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+env/teardown:
+	./bin/envteardown.sh
+
+dev:
+	make env/start
+	./bin/dev.sh
+
+test:
+	docker-compose -f docker-compose.test.yml --project-name nimble-survey-web-test up -d db redis
+	bundle exec rspec $(wordlist 2,$(words $(MAKECMDGOALS)),$(MAKECMDGOALS))
```
[the remark this answers is not in the document]
Comment (hoangmirs) [reply Makefile:30]: Good catch, we don't need to use the separated containers. Fixed in https://github.com/nimblehq/nimble-survey-web/pull/84/commits/4f2d77b729e125058c381d7a95ebec925f3e4767 and I also addressed all other comments

**Pull request merged**